### ISSUE TITLE
PERF: df.equals() does not check same shape to improve speed

### ISSUE THREAD

**Issue by DavidSiretMarques** [pandas-dev/pandas#53945]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

I was looking in stackoverflow and found [this question](https://stackoverflow.com/questions/76587013/pandas-dataframe-equals-why-not-first-checking-equal-shapes) where they asked why does pd.equals() not check for the shape as a first measurement of equality (if they have different shape, they're not equal). I checked and there seems to be no explanation, so I opened this issue. Here's the MWE from the question, I run it in my computer and it takes around 10 seconds more without the shape check.

```
import pandas as pd
import numpy as np
import itertools
import time
import random


def flag_duplicate_v1(dfs) -> None:
    for pairs in itertools.permutations(dfs.keys(), r=2):
        df1 = dfs[pairs[0]]
        df2 = dfs[pairs[1]]
        if type(df1) == str or type(df2) == str:
            # df1 or df2 has been marked duplicate in this loop, skip
            # comparing to remaining non-duplicate is enough
            continue

        if df1.equals(df2):
            print(f"df1 {pairs[0]} and df2 {pairs[1]} are duplicates")
            dfs[pairs[1]] = "DUPLICATE"
    return dfs


def flag_duplicate_v2(dfs) -> None:
    for pairs in itertools.permutations(dfs.keys(), r=2):
        df1 = dfs[pairs[0]]
        df2 = dfs[pairs[1]]
        if type(df1) == str or type(df2) == str:
            # df1 or df2 has been marked duplicate in this loop, skip
            # comparing to remaining non-duplicate is enough
            continue

        if df1.shape == df2.shape:
            if df1.equals(df2):
                print(f"df1 {pairs[0]} and df2 {pairs[1]} are duplicates")
                dfs[pairs[1]] = "DUPLICATE"
        # else not same shape so not duplicates
    return dfs


if __name__ == "__main__":
    # create number_df random dataframes
    dfs = {}
    number_df = 500
    for i in range(number_df):
        length = random.randint(1, 10)
        df = pd.DataFrame(
            {
                "A": np.random.randn(length),
                "B": np.random.randint(2, 10, length),
                "C": [
                    np.random.choice(["rand_1", "rand_2", "rand_4", "rand_6"])
                    for i in range(length)
                ],
            }
        )
        dfs[i] = df
    # add a duplicate
    dfs[number_df] = dfs[random.randint(1, number_df - 1)]

    tic_full_process = time.perf_counter()
    dfs_with_duplicates_v1 = flag_duplicate_v1(dfs)
    print(f"V1 : {time.perf_counter() - tic_full_process:0.1f}")

    tic_full_process = time.perf_counter()
    dfs_with_duplicates_v2 = flag_duplicate_v2(dfs)
    print(f"V2 : {time.perf_counter() - tic_full_process:0.1f}")

    if dfs_with_duplicates_v1 == dfs_with_duplicates_v2:
        print("SAME RESULT")
    else:
        print("DIFFERENT RESULT")
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 2e218d10984e9919f0296931d92ea851c6a6faf5
python           : 3.9.16.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19045
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 9, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : es_ES.cp1252

pandas           : 1.5.3
numpy            : 1.23.5
pytz             : 2022.7
dateutil         : 2.8.2
setuptools       : 66.0.0
pip              : 23.0.1
Cython           : None
pytest           : 7.1.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.9.2
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.12.0
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : 1.3.5
brotli           :
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.7.1
numba            : None
numexpr          : 2.8.4
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.9.3
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : 0.19.0
tzdata           : None
</details>

**Comment by jbrockmendel** [comment]
DataManager.equals has a check on axes that should be pretty similar to a check on shapes.  im surprised adding an explicit shape check makes such a big difference

**Comment by DavidSiretMarques** [comment]
I thougth it was weird it didn't check the shape...

So it does, but it's not fast... Interesting

**Comment by rouille** [comment]
> DataManager.equals has a check on axes that should be pretty similar to a check on shapes. im surprised adding an explicit shape check makes such a big difference

It seems that the check on axes is not enough. The below 2 data frames have different shapes:
```
>>> import pandas as pd
>>> df1 = pd.DataFrame({"A": [1,2,3], "B":[10,11,12]})
>>> df2 = pd.DataFrame({"A": [1,2,3,4], "B":[10,11,12,13]})
>>> df1.shape
(3, 2)
>>> df2.shape
(4, 2)
```
but the check on `axes` as implemented [here](https://github.com/pandas-dev/pandas/blob/7e0bcf1681a66f9bdf373ea549b2104e1a72184c/pandas/core/internals/base.py#L127), only checks the length, which is the same for the 2 data frames:
```
>>> len(df1.axes)
2
>>> len(df2.axes)
2
```  

I have a PR up (#53959) that check the shape in the `equals` method of the `NDFrame` class in the `pandas.core.generic` module

**Comment by jbrockmendel** [comment]
```
        self_axes, other_axes = self.axes, other.axes
        if len(self_axes) != len(other_axes):
            return False
        if not all(ax1.equals(ax2) for ax1, ax2 in zip(self_axes, other_axes)):
```

It's the last line here that is checking the shapes.

**Comment by rouille** [comment]
> ```
>         self_axes, other_axes = self.axes, other.axes
>         if len(self_axes) != len(other_axes):
>             return False
>         if not all(ax1.equals(ax2) for ax1, ax2 in zip(self_axes, other_axes)):
> ```
> 
> It's the last line here that is checking the shapes.

You are right. I am not sure why I did not see that in the first place, sorry for overlooking. Quick question, from a design point of view, would it make sense to move these checks in the `equals` method of the [`NDFrame` class](https://github.com/pandas-dev/pandas/blob/1186ee0080a43a08c3431a5bfa86898f6ec3b54d/pandas/core/generic.py#L228)?

**Comment by phofl** [comment]
Why? Manager is appropriate

**Comment by rouille** [comment]
> Why? Manager is appropriate

I thought it would make sense since both the `Series` and `DataFrame` classes inherit from `NDFRame`. The `equals` method there could hold the common checks on type and size.

**Comment by jbrockmendel** [comment]
The check exists and is in the appropriate place.  Closing.
